File: native-ios/RandomTimer/Sources/Services/RuntimeConfigurationService.swift
import CryptoKit
import Foundation

struct RuntimeConfigurationSnapshot: Equatable {
    let defaultConfig: TimerConfig
    let configSource: String
    let configVersion: String
    let experiments: [String: String]

    static let bundled = RuntimeConfigurationSnapshot(
        defaultConfig: .default,
        configSource: "bundled",
        configVersion: "bundled",
        experiments: [:]
    )

    var analyticsProperties: [String: Any] {
        var properties: [String: Any] = [
            AnalyticsProperties.runtimeConfigSource: configSource,
            AnalyticsProperties.runtimeConfigVersion: configVersion,
        ]
        experiments.forEach { key, value in
            properties["experiment_\(key)"] = value
        }
        return properties
    }
}

struct RuntimeConfigurationPayload: Decodable, Equatable {
    let configVersion: String
    let defaultTimerConfig: TimerConfig
    let experiments: [RuntimeExperimentDefinition]

    func toSnapshot(distinctId: String) -> RuntimeConfigurationSnapshot {
        RuntimeConfigurationSnapshot(
            defaultConfig: defaultTimerConfig,
            configSource: "insforge_storage",
            configVersion: configVersion,
            experiments: RuntimeExperimentAssigner.assign(distinctId: distinctId, experiments: experiments)
        )
    }
}

struct RuntimeExperimentDefinition: Decodable, Equatable {
    let key: String
    let variants: [RuntimeExperimentVariant]
}

struct RuntimeExperimentVariant: Decodable, Equatable {
    let key: String
    let rolloutPercent: Int
}

enum RuntimeExperimentAssigner {
    static func assign(
        distinctId: String,
        experiments: [RuntimeExperimentDefinition]
    ) -> [String: String] {
        Dictionary(
            uniqueKeysWithValues: experiments.compactMap { definition in
                guard let variant = chooseVariant(distinctId: distinctId, definition: definition) else {
                    return nil
                }
                return (definition.key, variant)
            }
        )
    }

    private static func chooseVariant(
        distinctId: String,
        definition: RuntimeExperimentDefinition
    ) -> String? {
        guard definition.variants.isEmpty == false else { return nil }

        let bucket = bucketFor(seed: "\(definition.key):\(distinctId)")
        var cumulative = 0
        for variant in definition.variants {
            cumulative += max(0, min(100, variant.rolloutPercent))
            if bucket < cumulative {
                return variant.key
            }
        }
        return nil
    }

    private static func bucketFor(seed: String) -> Int {
        let digest = SHA256.hash(data: Data(seed.utf8))
        let value = digest.prefix(8).reduce(UInt64(0)) { partial, byte in
            (partial << 8) | UInt64(byte)
        }
        return Int(value % 100)
    }
}

@MainActor
final class RuntimeConfigurationService {
    private(set) var snapshot: RuntimeConfigurationSnapshot = .bundled

    private let session: URLSession
    private var didRefresh = false

    init(session: URLSession = .shared) {
        self.session = session
    }

    func refreshIfNeeded(distinctId: String?) async {
        guard didRefresh == false else { return }
        didRefresh = true
        guard
            let distinctId,
            distinctId.isEmpty == false,
            let objectURL,
            apiKey.isEmpty == false
        else {
            return
        }

        do {
            let payload = try await fetchPayload(from: objectURL, apiKey: apiKey)
            snapshot = payload.toSnapshot(distinctId: distinctId)
        } catch {
            // Optional runtime config. Fall back silently to bundled defaults.
        }
    }

    func applyPayloadForTesting(_ payload: RuntimeConfigurationPayload, distinctId: String) {
        snapshot = payload.toSnapshot(distinctId: distinctId)
    }

    private var objectURL: URL? {
        guard
            let raw = Bundle.main.object(forInfoDictionaryKey: "INSFORGE_API_BASE_URL") as? String,
            raw.isEmpty == false
        else {
            return nil
        }
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines).trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        return URL(string: "\(trimmed)/api/storage/buckets/training_assets/objects/runtime/mobile-runtime-config.json")
    }

    private var apiKey: String {
        Bundle.main.object(forInfoDictionaryKey: "INSFORGE_API_KEY") as? String ?? ""
    }

    private func fetchPayload(from objectURL: URL, apiKey: String) async throws -> RuntimeConfigurationPayload {
        var request = URLRequest(url: objectURL)
        request.httpMethod = "GET"
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.timeoutInterval = 5

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw URLError(.badServerResponse)
        }
        return try JSONDecoder().decode(RuntimeConfigurationPayload.self, from: data)
    }
}

File: native-android/app/src/main/java/com/iganapolsky/randomtimer/runtime/RuntimeConfigurationService.kt
package com.iganapolsky.randomtimer.runtime

import com.iganapolsky.randomtimer.BuildConfig
import com.iganapolsky.randomtimer.domain.model.SoundType
import com.iganapolsky.randomtimer.domain.model.TimerConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.absoluteValue

data class RuntimeConfigurationSnapshot(
    val defaultConfig: TimerConfig = TimerConfig.DEFAULT,
    val configSource: String = "bundled",
    val configVersion: String = "bundled",
    val experiments: Map<String, String> = emptyMap(),
) {
    fun analyticsProperties(): Map<String, Any> =
        buildMap {
            put("runtime_config_source", configSource)
            put("runtime_config_version", configVersion)
            experiments.forEach { (key, value) ->
                put("experiment_$key", value)
            }
        }
}

data class RuntimeConfigurationPayload(
    val configVersion: String,
    val defaultTimerConfig: TimerConfig,
    val experiments: List<RuntimeExperimentDefinition>,
) {
    fun toSnapshot(distinctId: String): RuntimeConfigurationSnapshot =
        RuntimeConfigurationSnapshot(
            defaultConfig = defaultTimerConfig,
            configSource = "insforge_storage",
            configVersion = configVersion,
            experiments = RuntimeExperimentAssigner.assign(distinctId, experiments),
        )
}

data class RuntimeExperimentDefinition(
    val key: String,
    val variants: List<RuntimeExperimentVariant>,
)

data class RuntimeExperimentVariant(
    val key: String,
    val rolloutPercent: Int,
)

internal object RuntimeExperimentAssigner {
    fun assign(
        distinctId: String,
        experiments: List<RuntimeExperimentDefinition>,
    ): Map<String, String> =
        experiments
            .mapNotNull { definition ->
                chooseVariant(distinctId, definition)?.let { definition.key to it }
            }.toMap()

    private fun chooseVariant(
        distinctId: String,
        definition: RuntimeExperimentDefinition,
    ): String? {
        if (definition.variants.isEmpty()) return null

        val bucket = bucketFor("${definition.key}:$distinctId")
        var cumulative = 0
        definition.variants.forEach { variant ->
            val rollout = variant.rolloutPercent.coerceIn(0, 100)
            cumulative += rollout
            if (bucket < cumulative) {
                return variant.key
            }
        }
        return null
    }

    private fun bucketFor(seed: String): Int {
        val digest = MessageDigest.getInstance("SHA-256").digest(seed.toByteArray())
        val raw =
            digest.take(8).fold(0L) { acc, byte ->
                (acc shl 8) or (byte.toLong() and 0xff)
            }
        return (raw.absoluteValue % 100).toInt()
    }
}

@Singleton
class RuntimeConfigurationService
    @Inject
    constructor() {
        private val _snapshot = MutableStateFlow(RuntimeConfigurationSnapshot())
        val snapshot: StateFlow<RuntimeConfigurationSnapshot> = _snapshot.asStateFlow()

        suspend fun refresh(distinctId: String?) {
            val baseUrl = BuildConfig.INSFORGE_API_BASE_URL.trim().removeSuffix("/")
            val apiKey = BuildConfig.INSFORGE_API_KEY.trim()
            if (distinctId.isNullOrBlank() || baseUrl.isBlank() || apiKey.isBlank()) {
                return
            }

            val payload =
                runCatching {
                    fetchPayload(
                        objectUrl = "$baseUrl/api/storage/buckets/training_assets/objects/runtime/mobile-runtime-config.json",
                        apiKey = apiKey,
                    )
                }.getOrNull() ?: return

            _snapshot.value = payload.toSnapshot(distinctId)
        }

        internal fun applyPayloadForTesting(
            payload: RuntimeConfigurationPayload,
            distinctId: String,
        ) {
            _snapshot.value = payload.toSnapshot(distinctId)
        }

        private suspend fun fetchPayload(
            objectUrl: String,
            apiKey: String,
        ): RuntimeConfigurationPayload =
            withContext(Dispatchers.IO) {
                val connection =
                    (URL(objectUrl).openConnection() as HttpURLConnection).apply {
                        requestMethod = "GET"
                        setRequestProperty("x-api-key", apiKey)
                        setRequestProperty("accept", "application/json")
                        connectTimeout = 5000
                        readTimeout = 5000
                        instanceFollowRedirects = true
                    }
                try {
                    val code = connection.responseCode
                    require(code in 200..299) { "Runtime config fetch failed with HTTP $code" }
                    val body = connection.inputStream.bufferedReader().use { it.readText() }
                    parsePayload(body)
                } finally {
                    connection.disconnect()
                }
            }

        companion object {
            internal fun parsePayload(json: String): RuntimeConfigurationPayload {
                val root = JSONObject(json)
                val configVersion = root.optString("configVersion").ifBlank { "unknown" }
                val timer = root.optJSONObject("defaultTimerConfig") ?: JSONObject()
                val experiments = root.optJSONArray("experiments") ?: JSONArray()

                return RuntimeConfigurationPayload(
                    configVersion = configVersion,
                    defaultTimerConfig =
                        TimerConfig(
                            minSeconds = timer.optInt("minSeconds", TimerConfig.DEFAULT.minSeconds),
                            maxSeconds = timer.optInt("maxSeconds", TimerConfig.DEFAULT.maxSeconds),
                            alarmDuration = timer.optInt("alarmDuration", TimerConfig.DEFAULT.alarmDuration),
                            hiddenMode = timer.optBoolean("hiddenMode", TimerConfig.DEFAULT.hiddenMode),
                            repeatEnabled = timer.optBoolean("repeatEnabled", TimerConfig.DEFAULT.repeatEnabled),
                            soundType = parseSoundType(timer.optString("soundType")),
                            volume = timer.optDouble("volume", TimerConfig.DEFAULT.volume.toDouble()).toFloat(),
                            vibrationEnabled = timer.optBoolean("vibrationEnabled", TimerConfig.DEFAULT.vibrationEnabled),
                        ),
                    experiments =
                        buildList {
                            for (index in 0 until experiments.length()) {
                                val experiment = experiments.optJSONObject(index) ?: continue
                                val key = experiment.optString("key")
                                if (key.isBlank()) continue
                                val variantsJson = experiment.optJSONArray("variants") ?: JSONArray()
                                val variants =
                                    buildList {
                                        for (variantIndex in 0 until variantsJson.length()) {
                                            val variant = variantsJson.optJSONObject(variantIndex) ?: continue
                                            val variantKey = variant.optString("key")
                                            if (variantKey.isBlank()) continue
                                            add(
                                                RuntimeExperimentVariant(
                                                    key = variantKey,
                                                    rolloutPercent = variant.optInt("rolloutPercent", 0),
                                                ),
                                            )
                                        }
                                    }
                                add(RuntimeExperimentDefinition(key = key, variants = variants))
                            }
                        },
                )
            }

            private fun parseSoundType(raw: String?): SoundType =
                runCatching {
                    SoundType.valueOf(raw.orEmpty().trim().uppercase())
                }.getOrDefault(TimerConfig.DEFAULT.soundType)
        }
    }
